[6기 Kwonkyu] TodoList with Ajax

Kwonkyu opened the pull request:
# 🚀미션 - Todo List for Team!
## 🎯 요구사항
- [X] User 추가하기
- [X] User의 투두리스트 불러오기
- [X] User 삭제하기
- [X] todoItem 추가하기
- [X] todoItem 불러오기
- [X] todoItem complete하기
- [X] todoItem 삭제하기
- [X] todoItem contents 내용 수정하기
## 🎯🎯 심화 요구사항
- [x] 데이터를 불러오기전 로딩바를 이용해, 사용자가 데이터가 불러와지고 있다는 것을 보여줍니다.
- [X] fetch api 사용하는 부분을 async await을 사용하여 리팩토링합니다.
- [X] github issue에서 라벨을 붙이는 것처럼, 우선순위에 따라서 badge를 추가합니다.
- [X] ES6 impot & export를 이용해 자바스크립트 파일을 리팩토링합니다.
## 🕵️‍♂️ 제약사항
- [X] User의 이름은 최소 2글자 이상이어야 합니다.
# 후기
비동기를 활용한 애플리케이션을 작성하는 경험은 처음이라 좀 어려웠네요. 어떻게 하면 더 효율적으로 짤 수 있을까 고민했지만 경험이 부족해서 뚜렷한 개선방안이 생각나지 않아 좀 일찍 코드 리뷰를 요청드리게 되었습니다. 많은 지적 부탁드립니다!

Kwonkyu committed:
implemented feature

Kwonkyu committed:
added comments. disabled some input fields when async with server.

Kwonkyu committed:
- fixed disabled attribute when input is duplicated.
- removed deprecated innerText and applied textContent.
- to check the element has 'id' attribute, use 'hasAttribute()' to get
boolean value!

Kwonkyu committed:
applied lint(prettify). implemented loading animation for each todo element

yujo11 commented on js/common.js:
```diff
@@ -0,0 +1,28 @@
+
+// 할 일이 몇 개 있는지 출력하는 텍스트(총 n 개)를 업데이트하는 로직.
+export function updateCountText() {
+  const todoListCountText = document.querySelector("span.todo-count strong");
```
```suggestion
  const todoListCountText = document.querySelector(".todo-count > strong");
```
querySelector를 사용하시면 위와 같은 코드로도 파싱이 가능합니다!

Kwonkyu replied:
CSS 셀렉터에 좀 더 익숙해지면 querySelector를 사용하는데 도움이 많이 되겠군요! 감사합니다.

yujo11 commented on js/filterview.js:
```diff
@@ -0,0 +1,66 @@
+  // 현재 선택된 필터에서 selected 클래스 제거.
+  selectedFilter.classList.remove("selected");
+  // 이후 현재 선택된 필터에 따라 선택된 필터를 저장하는 변수 업데이트, 해당하는 할 일 display 속성 변경.
+  if (target.classList.contains("all")) {
+    selectedFilter = filterAll;
+    for (let index = 0; index < todoListCount; index++) {
+      todoElements[index].style.display = "";
+    }
+  } else if (target.classList.contains("active")) {
+    selectedFilter = filterActive;
+    for (let index = 0; index < todoListCount; index++) {
+      todoElements[index].style.display =
+        todoElements[index]
+          .querySelector("div.view input")
+          .getAttribute("checked") === null
+          ? ""
+          : "none";
+    }
+  } else if (target.classList.contains("completed")) {
+    selectedFilter = filterCompleted;
+    for (let index = 0; index < todoListCount; index++) {
+      todoElements[index].style.display =
+        todoElements[index]
+          .querySelector("div.view input")
+          .getAttribute("checked") === null
+          ? "none"
+          : "";
+    }
+  }
```
이 부분에서 조건문과 반복문이 여러번 중첩되어서 가독성이 조금 떨어진다고 느껴지네요ㅠㅠ 반복문만이라도 외부에 별도의 함수로 분리해주시면 보다 가독성 좋은 코드가 될거 같습니다!

Kwonkyu replied:
코드를 작성할 때는 생각이 안났는데 지금 다시보니 반복되는 부분을 충분히 빼서 간결하게 작성할 수 있을것 같네요! 함수로 최대한 분리해보겠습니다.

review#1 커밋에서 반영되었습니다. 감사합니다.

yujo11 commented on js/filterview.js:
```diff
@@ -0,0 +1,66 @@
+}
+
+export function applySelectedFilter() {
+  getSelectedFilter().dispatchEvent(new Event("click", { bubbles: true }));
```
```dispatchEvent()```라는 메서드도 있군요! 처음 알았네요!👍👍👍 

혹시 이 부분에서 어떤 이유로 ```dispatchEvent()```를 사용하신지 알 수 있을까요?

Kwonkyu replied:
현재 선택된 필터 버튼을 클릭하는 모의 이벤트를 발생시켜서 새로 추가되거나 상태가 변경된 할 일들이 할 일 목록에 나타나거나 나타나지 않도록 하기 위해 사용했습니다!

예를 들어 '해야할 일' 필터를 적용하고 있을 때 어떤 할 일의 체크박스를 클릭해서 완료 상태로 전환하면 '해야할 일' 필터에서는 보이지 않아야 하기 때문에 이를 숨기기 위해 다시 필터를 클릭하는 이벤트를 발생시킨 것입니다.

필터를 클릭했을 때는 필터를 담고 있는 엘리먼트(ul.filters)에서 이벤트 위임을 통해 할 일들의 display 속성을 조절하여 할 일들을 나타내거나 숨기고 있습니다.

yujo11 commented on js/users.js:
```diff
@@ -0,0 +1,146 @@
+        .catch(error => alert(error))
+    return updatedTodoElement
+}
+
```
```fetch``` 사용하는 부분을 굉장히 깔끔하게 작성하셨군요!👍👍👍 다만 메서드 별로 headers, body 등 중복되는 내용은 별도로 분리해주시면 더 깔끔한 코드가 될거 같습니다!

그리고 한 파일(모듈)에서 내보내는(export) 함수나 변수가 많을 경우 하나로 묶어서 export하는 방법도 추천드립니다!
```js
export const API = {
  addUser,
  deleteUser,
  ...
}
```

이런 식으로 export하게 되면 import 해서 사용할 때 아래와 같이 사용할 수 있어서 보다 코드가 깔끔해질거 같습니다!

```js
import { API } from '../user.js'

const user = API.addUser();
```

Kwonkyu replied:
중복되는 내용을 어떻게 분리할까 생각해보다가 함수를 적용해서 개선할 수 있었습니다. 감사합니다!
하나로 묶어서 export 하는 방법이 있는 줄은 몰랐네요. API 객체를 만들어서 메서드로 해당 함수들을 등록해주면 import 하는 쪽에서 이 객체를 불러와서 객체의 메서드를 사용하는 방법이라고 이해하면 될까요?

review#2, review#3 커밋에서 반영되었습니다. 감사합니다.

yujo11 replied:
넵넵 맞습니다!👍👍👍

yujo11 commented on js/todolistview.js:
```diff
@@ -0,0 +1,394 @@
+  if (key === "Escape") {
+    todoElement.classList.toggle("editing");
+  } else if (key === "Enter") {
+    const newTodoText = target.value.trimStart().trimEnd();
```
```suggestion
    const newTodoText = target.value.trim();
```
```trim()```메서드를 사용하게 되면 문자열의 앞, 뒤 공백을 모두 제거해주기 때문에 ```trimStart()```와 ```trimEnd()```를 모두 사용하는 것보다 ```trim()``` 단일 메서드를 사용하는게 더 좋을거 같습니다!

Kwonkyu replied:
훨씬 편리하네요. 감사합니다!

review#4 커밋에서 반영되었습니다.

yujo11 commented on js/todolistview.js:
```diff
@@ -0,0 +1,394 @@
+}
+
+// 할 일을 클릭했을때 이벤트 위임.
+function onTodoElementClicked(event) {
+  if (!event.target) {
+    return;
+  }
+
+  if (
+    event.target.nodeName === "INPUT" &&
+    event.target.classList.contains("toggle")
+  ) {
+    toggleTodoElementStatus(event);
+  } else if (event.target.nodeName === "BUTTON") {
+    removeCurrentTodoElement(event);
+  }
+}
```
이런식으로 분기를 주는 방법도 좋네요!👍👍👍

어떤 함수에서는 구조분해 할당을 통해 ```{ target }```을 받아서 사용하셨는데 이 함수에서는 ```evnet```를 통채로 받아오신 이유가 있을까요?

Kwonkyu replied:
구현과정 중에 event 객체가 필요한 게 있었다가 나중에 필요없어지면서 남아있던 흔적인 것 같습니다. preventDefault() 함수를 사용하느라 event 객체가 필요한 함수도 있지만 필요없는 함수들도 몇개 남아있었는데 못보고 지나칠뻔했네요! 감사합니다.

review#5 커밋에서 반영되었습니다.

yujo11 reviewed (commented):
코드 잘 봤습니다! 지난 주보다 코드가 훨씬 깔끔하고 보기 좋아진거 같습니다!!👍👍👍👍

다양한 예외사항을 산정하면서 안전한 코드를 작성하기 위해 신경쓰신 부분들도 정말 멋졌습니다!!👍👍👍

다만 하나의 파일(todolistview)에서 굉장히 많은 일들을 수행하고 있는데 이 부분을 모듈화 시켜주시면 가독성이 더욱 좋아질거 같습니다! 

고생하셨습니다!!👍👍👍

JunseokPark13 commented on js/todolistview.js:
```diff
@@ -0,0 +1,394 @@
+    selectedPriority == 0 ? "NONE" : selectedPriority == 1 ? "FIRST" : "SECOND"
+  );
+  // 서버측 결과에 따라 스타일, 선택 항목 적용.
+  switch (updatedTodoElement.priority) {
+    case "FIRST":
+      target.value = 1;
+      target.classList.remove("secondary");
+      target.classList.add("primary");
+      break;
+    case "SECOND":
+      target.value = 2;
+      target.classList.remove("primary");
+      target.classList.add("secondary");
+      break;
+    case "NONE":
+      target.value = 0;
+      target.classList.remove("primary");
+      target.classList.remove("secondary");
+  }
```
classList 에서 remove 메서드를 수행할 때, 없는 클래스를 지우더라도 에러가 발생하지 않는다고 합니다. 
또한 하나의 remove 메서드에 여러 클래스명을 넣어서 동시에 지우는 것도 가능합니다!
primary 와 secondary 를 모두 삭제한 뒤 케이스에 맞게 필요한 클래스만 추가하는 방식으로 작성하시면 더 코드가 간결해질 것 같습니다! 

```suggestion
  target.classList.remove("primary", "secondary");
  switch (updatedTodoElement.priority) {
    case "FIRST":
      target.value = 1;
      target.classList.add("primary");
      break;
    case "SECOND":
      target.value = 2;
      target.classList.add("secondary");
      break;
    case "NONE":
      target.value = 0;
  }
```

Kwonkyu replied:
여러개를 지울 수 있는지는 몰랐네요. 훨씬 더 좋은 방법인것 같습니다. 감사합니다!

Kwonkyu committed:
Update js/common.js

selector "span.todo-count strong" to ".todo-count > strong". ">" selector means parent-child relation.

Co-authored-by: JO YUN HO <bedro27@gmail.com>

Kwonkyu committed:
review#1: changed structure of filter applying logic.

Kwonkyu committed:
review#2: implemented function to produce fetch request's option.

Kwonkyu committed:
review#3: optimized import, export statement.

Kwonkyu committed:
review#4: changed trimStart().trimEnd() to trim() call.

Kwonkyu committed:
review#5: destructuring event parameters.

Kwonkyu committed:
Update js/todolistview.js

optimized priority changed logic.

Co-authored-by: Bandito <49611158+Suppplier@users.noreply.github.com>

imakerjun commented on js/application.js:
```diff
@@ -0,0 +1,11 @@
+import {initTodos} from './todolistview.js'
+import {initFilters} from './filterview.js'
+
+// 웹페이지 로드 시 init() 함수 실행
```
주석은 꼭 필요한 경우만 남겨주시는면 더 간결해지고, 다른 사람들과 협업하기 좋은 코드가 될 것 같아요  😄

Kwonkyu replied:
주석을 최대한 없애고 함수로 분리해서 함수명만으로 유추 가능한 코드를 작성해보도록 노력하겠습니다!

review#6 커밋에서 반영하였습니다.

imakerjun commented on js/todolistview.js:
```diff
@@ -0,0 +1,375 @@
+}
+
+async function onClickUserList({ target }) {
+  if (target && target.nodeName === "BUTTON") {
```
아래와 같이 작성해주시면 depth도 줄어들고, 코드 로직이 더 명확하게 잘 보일 것 같아요~!
```javascript
if (!target || target.nodeName !== "BUTTON") {
  return
}

if (target.hasAttribute("id")) { ... }
```

Kwonkyu replied:
코드의 depth를 줄이면 조금 길어질지는 몰라도 괄호 처리도 편하고 더 좋은점이 많은것 같습니다. 습관을 들여야겠네요👍

review#6 커밋에서 반영하였습니다.

imakerjun commented on js/todolistview.js:
```diff
@@ -0,0 +1,375 @@
+async function updateTodoEdit({ target, key }) {
+  const todoElement = target.closest("li");
+  if (key === "Escape") {
+    todoElement.classList.toggle("editing");
```
여기서 return을 바로 해주시면 key가 Escape인 경우 여기까지만 동작하고, 굳이 밑에 else if까지 비교하는 검증까지 내려가지 않을 것 같아요

Kwonkyu replied:
이것도 코드의 depth를 줄이는 개선방안이군요. 코드에 전체적으로 적용하도록 하겠습니다!

review#6 커밋에서 반영하였습니다.

imakerjun commented on js/filterview.js:
```diff
@@ -0,0 +1,72 @@
+  // 현재 선택된 필터에서 selected 클래스 제거.
+  selectedFilter.classList.remove("selected");
+  // 이후 현재 선택된 필터에 따라 선택된 필터를 저장하는 변수 업데이트, 해당하는 할 일 display 속성 변경.
+  if (target.classList.contains("all")) {
+    selectedFilter = filterAll;
+    applyTodoElementsFilter(todoElements, applyFilterAll);
+  } else if (target.classList.contains("active")) {
+    selectedFilter = filterActive;
+    applyTodoElementsFilter(todoElements, applyFilterActive);
+  } else if (target.classList.contains("completed")) {
+    selectedFilter = filterCompleted;
+    applyTodoElementsFilter(todoElements, applyFilterCompleted);
+  }
```
필터 비교 로직이 중복되고 있는데, 콜백함수로 해당 조건에 실행해줄 함수만 전달하고, 비교하는 부분은 따로 분리하면 어떨까요?
아래와 같이 오브젝트로 관리해서 사용할수도 있을 것 같아요!
```js
const filterHandler = {
    all: () => {},
    active: () => {},
    completed: () => {},
}
```

Kwonkyu replied:
얼마전에서야 자바스크립트에서 함수가 일급 객체라는 사실을 알 수 있었습니다. 콜백 함수를 활용하면서 함수의 크기를 줄이고 최대한 유연하게 코드를 작성해야겠습니다👍👍

review#7 커밋에서 반영되었습니다. 감사합니다.

imakerjun commented on js/todolistview.js:
```diff
@@ -0,0 +1,375 @@
+// 할 일 추가 시 실제로 HTML 요소를 그리는 함수
+function drawTodo({ _id, contents, priority, isCompleted }) {
+  if (
+    !_id === undefined ||
+    !contents === undefined ||
+    !priority === undefined ||
+    isCompleted === undefined
```
if문안에 조건이 길어진다면 함수로 따로 분리해서 비교하면 더 가독성이 좋을 것 같아요! 해당 부분은 `isValidTodo()`와 같이 구분한다면 조건문 로직의 의미도 명확해지고요!

Kwonkyu replied:
해당 조건을 검사하는 로직은 충분히 재사용될 수 있기 때문에 말씀대로 함수로 추출하는게 좋은 방법인것 같습니다. 감사합니다!👍

review#6 커밋에서 반영하였습니다.

imakerjun commented on js/todolistview.js:
```diff
@@ -0,0 +1,375 @@
+
+// 모든 할 일들 제거.
+async function onRemoveAllTodoClick(event) {
+  if (await UserAPI.deteleAllTodoElement(UserAPI.getActiveUserID())) {
+    clearTodoList();
+  } else {
+    // notify user that operation is failed.
+    alert("할 일을 제거하는 데 실패했습니다!");
+  }
```
`try ~ catch`로 변경하면 의미가 더 명확해질 것 같아요~! 
실패를 나타내는 것이니깐요 ㅎㅎ

Kwonkyu replied:
시도와 실패를 다룰때는 try-catch를 사용하는 습관을 들여야겠습니다. 감사합니다!

review#6 커밋에서 반영하였습니다.

imakerjun commented on js/users.js:
```diff
@@ -0,0 +1,166 @@
+}
+
+function requestOption(method) {
+  return { method: method };
```
key와 value가 같은 경우 `return { method }`로 축약해서 쓰는 것도 가능합니다!

Kwonkyu replied:
이건 정말 몰랐네요! 감사합니다.

imakerjun reviewed (changes requested):
Kwonkyu님 세부적인 부분까지 꼼꼼하게 처리하시려던 점이 돋보였습니다 👍🏻👍🏻👍🏻
여기서 한발자국 더 나아가기 위해서 한 함수나 메서드를 가능하면 작은 단위로 쪼개서, 한 함수가 한 기능만 하게끔 리팩토링 하는 연습만 조금 더 해주시면 훨씬 더 좋은 코드가 될 것 같아요 😄

Kwonkyu committed:
review#6: decreased depth of codes and removed unnecessary comments.

Kwonkyu committed:
review#7: removed duplicated code and added more functions
